### PR TITLE
[Prover] Generalize handling of native methods in borrow checker

### PR DESCRIPTION
**Pull request by awelc**
## Motivation

The goal of these changes is to support other Move language flavors implementing alternative storage models (it resolves https://github.com/move-language/move/issues/660). The main idea is that alternative storage models may need support for implementing custom mutable borrows along the same lines as "core" Move's `borrow_mut` for vectors or tables works:
```
native public fun borrow_mut<Element>(v: &mut vector<Element>, i: u64): &mut Element;

public fun borrow_mut<K: copy + drop, V>(table: &mut Table<K, V>, key: K): &mut V {
   &mut borrow_box_mut<K, V, Box<V>>(table, key).val
}
```

These functions get a special treatment from the prover to artificially create an edge from their first parameter to their return value. Additionally, Operations on the values being borrowed from have to be explicitly defined in Boogie and referred to during bytecode translation.

This PR extends this mechanism to allow native functions defined outside of "core" Move to take advantage of the same custom support. 

### Have you read the [Contributing Guidelines on pull requests](https://github.com/move-language/move/blob/main/CONTRIBUTING.md#developer-workflow)?

Yes

### CODE REVIEW

File: language/move-prover/bytecode/src/borrow_analysis.rs [hunk 1]
```diff
@@ -419,15 +426,59 @@ impl FunctionTargetProcessor for BorrowAnalysisProcessor {
...
+                )
+            } else if fun_env.is_intrinsic_of(INTRINSIC_FUN_MAP_BORROW_MUT) {
+                (
+                    "GetTable".to_string(),
+                    "UpdateTable".to_string(),
+                    type_args[1].clone(),
+                )
+            } else {
+                let (r, u, i) = borrow_native_info.unwrap();
+                (r, u, type_args[i].clone())
```
Comment (emmazzz) [inline language/move-prover/bytecode/src/borrow_analysis.rs:479]: Very nice!

File: language/move-model/src/model.rs [hunk 2]
```diff
@@ -2933,7 +2934,7 @@ pub struct FunctionData {
 
     /// List of type argument names. Not in bytecode but obtained from AST.
     #[allow(unused)]
-    type_arg_names: Vec<Symbol>,
+    type_arg_names: Vec<(Symbol, Type)>,
```
Comment (emmazzz) [inline language/move-model/src/model.rs:2937]: Maybe this field should be renamed since it's no longer just names of the type args?

File: language/move-prover/bytecode/src/options.rs [hunk 3]
```diff
@@ -29,6 +29,45 @@ impl AutoTraceLevel {
...
+    /// Method's module name
+    pub mod_name: String,
+    /// Method's name
+    pub name: String,
+    /// Name of a read operation for a given type (in Boogie)
+    pub read_op: String,
+    /// Name of a Write operation for a given type (in Boogie)
+    pub write_op: String,
+    /// Method's type parameter index describing a give type
+    pub tparam_idx: usize,
```
Comment (emmazzz) [inline language/move-prover/bytecode/src/options.rs:48]: This seems to assume that the borrow type is one of the type parameters? What if the function has no type parameter or if we have type params `<K, V>` but the borrow type is `Wrapper<V>`? 
I'm fine with the way it is currently too if you think allowing those scenarios would be overgeneralization.

Comment (awelc) [reply language/move-prover/bytecode/src/options.rs:48]: I agree with your assessment of the limitation, but I am not sure how much it makes sense to generalize it at this point. I'd think that the current solution is already more general than the previous one which assumed that the required type is a type parameter of the argument (e.g., `Table<K,V>`) and it covers all the cases we need to cover right now.

If someone has a suggestion on how/why to generalize it further I am all ears, but perhaps we should consider it only if further need arises?

File: language/move-model/src/model.rs (hunk 2 above)
Comment (awelc) [reply language/move-model/src/model.rs]: Good call! Done!

File: language/move-prover/bytecode/src/options.rs (hunk 3 above)
Comment (awelc) [reply language/move-prover/bytecode/src/options.rs]: After pondering this for a bit longer I realized that we probably don't need the type info at all as it was used only to figure out which "access" functions to use in the bytecode translator.

File: language/move-prover/bytecode/src/stackless_bytecode.rs [hunk 4]
```diff
@@ -298,7 +298,7 @@ pub enum BorrowEdge {
     /// Field borrow with static offset.
     Field(QualifiedInstId<StructId>, usize),
     /// Vector borrow with dynamic index.
-    Index,
+    Index((String, String)),
```
Comment (meng-xu-cs) [inline language/move-prover/bytecode/src/stackless_bytecode.rs:301]: I would recommend against bringing in things that are only know to Boogie here. Instead try to capture the borrow semantics with a higher-level of abstraction (I am proving a solution below as well).

The reason we have `Index` here as a `BorrowEdge` is to mark a conceptual linkage with the SMT array theory. If you take a closer examination of the two things that are currently being modeled as `Index`, they are `Vector` borrows and `Table` borrows. Both resembles essentially the SMT array theory of `K -> V` mapping (yes, vector is just a table with `K = int`).

So, if you are re-purposing the `Index` one with an extended and custom semantics, I am really really really curious about what are the new type of mapping you try to model here.

Based on what I have learned from our prior discussions, here is what I think you are trying to do: you are tying to model a `K -> Any` mapping to simulate heterogeneous maps that can grow dynamically. And perhaps, the `K` can be further narrowed down to `int` or `str` depending on how you mark the field identifier.

This is indeed a new functionality and justifies this new design which is beautifully implemented in this PR (but note: this is a different issue from #660). In fact, I believe that this `Heterogeneous Map` feature is a generic feature that warrants more than just a custom Boogie treatment.

The solution I am proposing to remove this `(String, String)` part in `Index(..)` is to use an enum:

```rust
pub enum ContainerKind {
    /// SMT array theory: int -> E (E is known)
    /// --> already have
    Vector,
    /// SMT array theory: K -> V (both K, V are known)
    /// --> already have
    KVMap,
    /// SMT array theory: K -> Any (K is known, Any can be any type, i.e., dynamic typing)
    /// --> I guess this is the new type of map you are envisioning
    AnyMap: 
}
```

And hence, `BorrowMap::Index(ContainerKind)`.

In `boogie-backend/src/bytecode_translator.rs`, we can further match this `ContainerKind` to roll-out the correct Boogie functions to for reference reading and updating.

However, this `Heterogeneous Map` thing is just a long shot guess and please correct me if this is wrong.

Comment (emmazzz) [reply language/move-prover/bytecode/src/stackless_bytecode.rs:301]: I believe the two `String`s in `Index((String, String))` are for the boogie function names of the custom read_aggregate and update_aggregate. It is populated at `native_annotation` in borrow_analysis.rs. So for vector index, these two strings will be "ReadVec" and "UpdateVec". For custom indices, these two strings are provided through the `BorrowNative` option. 

The first paragraph of your comment still makes sense though. Where do you think is a better place to put these custom function names? @meng-xu-cs 

Using an enum might be a good idea too if the kind of mapping we'll be doing is vastly different from what we currently have. This part hasn't really been designed yet afaik. Right now we are just trying to get Sui Move prover translation (not trying to prove any specs yet) working without errors.

Comment (emmazzz) [reply language/move-prover/bytecode/src/stackless_bytecode.rs:301]: Chatted with @meng-xu-cs earlier this week about this. IIUC the reason why we currently special case for table and vector modules is because they correspond to the two Boogie constructs we have. If the customization we want to add can be represented or modeled using the current constructs, we don't need to provide new aggregate functions. Instead, we should provide custom mapping with information specifying "this custom type should use table Boogie construct". @awelc do you think this is doable?

Comment (awelc) [reply language/move-prover/bytecode/src/stackless_bytecode.rs:301]: > we should provide custom mapping with information specifying "this custom type should use table Boogie construct". @awelc do you think this is doable?

I understand the argument better now, and I agree that this would be ideal, but I don't think it's possible (happy to be proven wrong, though!). The way I currently understand it (though admittedly my understanding is limited), table accesses are modeled by having per-table arrays storing relevant information:
```
// v is the SMT array holding the key-value assignment. e is an array which
// independently determines whether a key is valid or not. l is the length.
//
// Note that even though the program cannot reflect over existence of a key,
// we want the specification to be able to do this, so it can express
// verification conditions like "key has been inserted".
function {:constructor} Table<K, V>(v: [K]V, e: [K]bool, l: int): Table K V;
```

These are used in the aggregate functions to model the read/write operations:

```
function {:inline} GetTable<K,V>(t: Table K V, k: K): V {
    v#Table(t)[k]
}

function {:inline} UpdateTable<K,V>(t: Table K V, k: K, v: V): Table K V {
    Table(v#Table(t)[k := v], e#Table(t)[k := true], l#Table(t))
}
```

The extension proposed in this PR aims at supporting a different scenario. We need to model accessing a piece of data (more specifically a Sui object) that can have an arbitrary type. In fact, the first parameter to the native function that needs modeling is a mutable reference to the unique ID  (`UID`) of an object rather than the object itself:

```
public(friend) native fun borrow_child_object_mut<Child: key>(object: &mut UID, id: address): &mut Child;
```

The aggregate functions modeling read/write operations to an object represented by a given `UID` must tap directly into the (global) mapping representing Sui's storage model and I don't think this can be mapped to the existing array theory, and will require custom aggregate function definitions. Something like the following where we get parent object info from global storage based on its address and then gets its children from the appropriate mapping):

```
function {:inline} GetChild<V>(parent_id: $2_object_UID, child_addr: int): V {
    children#$Memory[uid_to_addr(parent_id)][child_addr]
}
```

Comment (meng-xu-cs) [reply language/move-prover/bytecode/src/stackless_bytecode.rs:301]: @awelc I sincerely ask you to read the comment again.

TLDR

```rust
pub enum ContainerKind {
    /// SMT array theory: int -> E (E is known)
    /// --> already have
    Vector,
    /// SMT array theory: K -> V (both K, V are known)
    /// --> already have
    KVMap,
    /// SMT array theory: K -> Any (K is known, Any can be any type, i.e., dynamic typing)
    /// --> I guess this is the new type of map you are envisioning
    AnyMap: 
}
```

```diff
pub enum BorrowEdge {
    /// Field borrow with static offset.
    Field(QualifiedInstId<StructId>, usize),
    /// Vector borrow with dynamic index.
-   Index,
+   Index(ContainerKind),
}
```

Comment (awelc) [reply language/move-prover/bytecode/src/stackless_bytecode.rs:301]: I did read the comment - my response was to the following comment by @emmazzz:

> If the customization we want to add can be represented or modeled using the current constructs, we don't need to provide new aggregate functions. 

Please also understand that I am doing my best here but I am also learning things as I go that may be super-obvious to everyone else...

I am super cool with introducing the new enum, but I think we still need the custom aggregate functions for `AnyMap`. Would the following (or something similar, like having a new struct as `AnyMap` variant's value instead of two strings representing names of the aggregate functions) make sense?

```
pub enum ContainerKind {
    /// SMT array theory: int -> E (E is known)
    /// --> already have
    Vector,
    /// SMT array theory: K -> V (both K, V are known)
    /// --> already have
    KVMap,
    /// SMT array theory: K -> Any (K is known, Any can be any type, i.e., dynamic typing)
    /// --> I guess this is the new type of map you are envisioning
    AnyMap(String, String)
}
```

I simply don't know how we can map what we need model to one of existing array theories (e.g., those contained in `table-array-theory.bpl ` or `vector-array-theory.bpl`). I think we need to provide our own model there and refer to it from custom aggregate functions"

Comment (meng-xu-cs) [reply language/move-prover/bytecode/src/stackless_bytecode.rs:301]: >  but I think we still need the custom aggregate functions for AnyMap.

We do. I am not fighting against this one.

I am only fighting against exposing the Boogie encoding as `(String, String)` in the bytecode. The Boogie encoding should only be known by things under `boogie-backend/src/*`.

Comment (awelc) [reply language/move-prover/bytecode/src/stackless_bytecode.rs]: Cool, thank you for the clarification! Now that I understand what the issue is, I have rewritten the code to avoid introducing Boogie-level aggregate names at the borrow analysis level. Instead, the `Custom` variant of the index edge encodes the name of the native function which needs special treatment and aggregate names are introduced as options to the backend rather than to the bytecode transformer. Hopefully this is acceptable

Another solution would be to create `Custom` variant without any values, and hardcode the names of aggregate functions for it somewhere in `boogie-backend/src/*` the same way that aggregate function names are hardcoded for vectors and tables?

This would satisfy our current need, but would be more limiting in that we could only have one native function that would get the custom treatment (as we would only have one set of hardcoded aggregate function names). I am hoping  for something more general, where we can have this custom treatment for multiple native functions (much like both table and vector aggregate functions are needed in core Move).

**Review by meng-xu-cs — APPROVED** [review]
Thanks for the changes. The new approach (of identifying a native borrow function by its name) look good to me. Accept to unblock and this is a good foundation even if we want future iterations.

*NOTE*: I noticed a stream of commits in this PR so I guess the best way to land is squashing first and then merge.
